### PR TITLE
[RFC] "Unrolled" representation of multi-task MVN

### PR DESCRIPTION
**Pull request by Balandat**
Currently, the interleaved/non-interleaved representation of MTMVN creates a bunch of headaches (in particular regarding indexing, scalarization, and performance). It would be nice to have a higher-level API that abstracts away from these details (see #1055).

In this RFC, the covariance matrix is represented as an "unrolled" tensor (e.g. `n x t x n x t`), instead of a `nt x nt` covariance matrix. The constructor accepts a covariance matrix of shape `batch_shape x shape`-dim tensor, where `shape` has either three or four elements, and can be any combination of "n" and "t" with at least one and at most two occurrences of both "n" and "t" (currently only n==2 is supported). Internally, the covariance is represented consistently as `n x t n x t` or `t x n x n` (case of cross-task independence) in order to simplify indexing / scalarization operations.

The main benefits here are: 
1) It makes indexing much easier (e.g. if we want the cross-point covariance at some task `j`, we can just index the internal covariance as `_covar[..., :, j, :, j]` and we are done. With the current representation (especially considering the interleaving option) this is a huge headache. Note that this RFC does not include any indexing functions)
2) We can represent MTMVNs with independent outputs much more efficiently, without necessarily having to do a ton of `LazyTensor` acrobatics. Importantly, this should mean that in this case sampling and computing log probs should be much faster. 
3) We can easily scalarize multi-output posteriors (in fact, this is related to indexing)

Currently, this is very early work, and so there are major limitations with this:
1) This PR assumes proper tensors throughout (no support for `LazyTensor` yet)
2) No caching of matrix decompositions (yet)
3) The class subclasses `MultivariateNormal` (and hence the torch  `MultivariateNormal`), but it doesn't make sure that it implements the full interface, probably a lot of stuff will just break.
4) No support for `covariance_matrix` and `lazy_covariance_matrix` interface. I am actually not sure that we'll need this though.

The notebook included in this PR demos the basic usage of this.

### CODE REVIEW

File: gpytorch/distributions/new_multitask_multivariate_normal.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            task_dim += len(batch_shape)
+        mean = mean.permute(*range(0, task_dim), *range(task_dim + 1, mean.ndim), task_dim)
+        cov = batch_mvn.covariance_matrix
+        covariance = cov.permute(*range(0, task_dim), *range(task_dim + 1, cov.ndim - 2), task_dim, -2, -1)
+        return cls(mean=mean, covariance=covariance, order="tnn")
+
+    @property
+    def event_shape(self) -> torch.Size:
+        return self._mean.shape[-2:]
+
```
Comment (gpleiss) [inline gpytorch/distributions/new_multitask_multivariate_normal.py]: We probably want a function somewhere that converts between representations.

File: gpytorch/distributions/new_multitask_multivariate_normal.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+class MultitaskMultivariateNormal(MultivariateNormal):
+    def __init__(self, mean: Tensor, covariance: Tensor, order: str = "ntnt") -> MultitaskMultivariateNormal:
+        """Construct a MultitaskMultivariateNormal from a stacked representation of its covariance matrix.
+
+        :param torch.tensor mean: A `batch_shape x n x t` tensor of means
+        :param torch.tensor covariance: A `batch_shape x shape`-dim tensor, where `shape` has either three
+            or four elements and can be any combination of "n" and "t" with at least one and at most two
+            occurrences of both "n" and "t" (currently only n==2 is supported)
+
+        Internally, the covariance is represented consistently as `n x t n x t` or `t x n x n` (case of
```
[the remark this answers is not in the document]
Comment (Balandat) [reply gpytorch/distributions/new_multitask_multivariate_normal.py]: That makes sense. The reason for me using `n x t x n x t` initially was that it's straightforward to get to a full `nt x nt` covariance matrix with a single `view` / `reshape` without doing much permutation acrobatics: `self._covar.view(*self._batch_shape, n * t, n * t)`. But I guess that's not all that important. 

Re using a `t x n x n` `KroneckerProductLazyTensor` - In this case I wouldn't actually use a `KroneckerProductLazyTensor`, we can just compute the log-likelihood in a batched fashion for each `n x n` matrix and sum them up. Using the `KroneckerProductLazyTensor` would probably create more overhead than it would help. For `t x t x n x n` it makes sense, of course.

**Comment by Balandat** [comment]
> Now as a next step - would it make sense to get this new MMVN implemented for non-lazy tensors, and focus on adding LTs later down the road?

I think for non-lazy tensors the functionality is essentially there, the rest is just some grunt work (especially caching the cholesky's / allowing for linearCG computations). So including support at least for `KroneckerProductLazyTensor` should be part of the next step.

Re the `covariance_matrix` and `lazy_covariance_matrix` APIs: Should we still allow this to be `interleaved` and not `interleaved`? It seems that this doesn't really matter much in this new representation and we can just agree on a convention. Any preferences?

File: gpytorch/distributions/new_multitask_multivariate_normal.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Balandat) [reply gpytorch/distributions/new_multitask_multivariate_normal.py]: Hmm, actually, thinking about this more, for Kronecker strucutre we can be a smarter rather than just expressing the `t x t x n x n` from the Kronecker representation (see e.g. the related https://github.com/cornellius-gp/gpytorch/pull/1086). So we'd actually want to just keep a `t x t` and a `n x n` tensor around, then we don't even need to reason about `t x t x n x n`.

**Comment by jacobrgardner** [comment]
@Balandat regarding `order=` naming, this might be a good opportunity to follow conventions set in PyTorch for named tensors? If we start building something like this in to LazyTensor, then we could just name the dimensions of the covariance matrix appropriately.

**Comment by Balandat** [comment]
Good call @jacobrgardner. I was an early adopter of NamedTensors, but upsterma development on them has been somewhat de-prioritized, so I didn't do much with them. But I think it makes a ton of sense here (and more generally for other stuff as well). 

I'll make sure to comply with this convention. In fact, we can even just accept NamedTensors as input args as well without requiring to specify the order.

**Comment by Balandat** [comment]
Also a lot of stuff is currently not supported with NamedTensors (e.g. permute), so while we can align the naming interface, we can't really do much with them until there is better support for general operations.

**Comment by Balandat** [comment]
Added a lazy version for the task-independent case. Not sure what level of generality we want / need to support for the lazies, at some point things become really clunky. One thing I will definitely add is a Kronecker version.

**Comment by Balandat** [comment]
@gpleiss, @jacobrgardner, @dme65 I finally got to get back to this.

To recall, I drafted a `MultioutputMultivariateNormal` distribution (subclasses `MultivariateNormal`, but for now completely separate from `MultitaskMultivariateNormal`) . This is an abstract API to a multi-output MVN that has is of the shape `batch_shape x n x m`, where `n` is the number of data points and `m` is the number of outputs. Internally, the representation can be whatever is most suitable to make computations / indexing more efficient / simpler. 

The main limitation right now is that I haven't really hooked this up to the lazies. But at least for the existing functionality I came up with something that allows for quite a nice indexing API (this all works, see the nb in the PR). Any thoughts on this? 

#### Multi-output MVN subsetting API:

```python
$ mtmvn
    
MultioutputMultivariateNormal(batch_shape=(2,), n=3, m=2)
```

```python
# select batch <- this returns an non-batched MOMVN
$ mtmvn[0]

MultioutputMultivariateNormal(n=3, m=2)
```

```python
# select range of batches <- this is a null-op
$ mtmvn[:]

MultioutputMultivariateNormal(batch_shape=(2,), n=3, m=2)
```

```python
# select data point <- returns a batched MVN (single-output)
$ mtmvn[..., 0, :]

MultivariateNormal(batch_shape=(2,), n=2)
```

```python
# select range of data points <- just subsets the data (n)
$mtmvn[..., :2, :]

MultioutputMultivariateNormal(batch_shape=(2,), n=2, m=2)
```

```python
# select single output <- returns a standard MVN
$ mtmvn[..., 0]

MultivariateNormal(batch_shape=(2,), n=3)
```

```python
# select range of outputs <- just subsets the tasks (m)
$ mtmvn[..., :2]

MultioutputMultivariateNormal(batch_shape=(2,), n=3, m=2)
```

```python
# mixed indexing select specific data point and output form specific batch
# This is a degenerate MVN with one data point
$ mtmvn[0, 1, 0]

MultivariateNormal(n=1)
```

```python
# mixed indexing but respect batch dims
$ mtmvn[..., 1, 0]

MultivariateNormal(batch_shape=(2,), n=1)
```

**Comment by Balandat** [comment]
@neerajprad you might be interested in this, too
